p2p/host/basic/basic_host.go

```diff
@@ -805,6 +808,22 @@ func (h *BasicHost) AllAddrs() []ma.Multiaddr {
 		finalAddrs = append(finalAddrs, resolved...)
 	}
 
+	// add autonat PublicAddr Consider the following scenario
+	// For example, it is deployed on a cloud server,
+	// it provides an elastic ip accessible to the public network,
+	// but not have an external network card,
+	// so net.InterfaceAddrs() not has the public ip
+	// The host can indeed be dialed ！！！
+	if h.AutoNat != nil {
+		ambientAutoNat, ok := h.AutoNat.(*autonat.AmbientAutoNAT)
+		if ok && ambientAutoNat != nil {
+			publicAddr, _ := ambientAutoNat.PublicAddr()
```

Comment:
we should probably add `PublicAddr` to the `AutoNAT` interface rather than needing to limit this check to the concrete implementation.